setup.py

```diff
@@ -94,7 +94,20 @@
 ]
 
 
+# Only add the constraints so that 'pip install qiskit-terra[all]' ignores bip-mapper on those
+# versions. If the user attempts to explicitly install 'pip install qiskit-terra[bip-mapper]', we
+# want to force pip to either degrade Terra or emit an error.
 bip_requirements = ["cplex", "docplex"]
+bip_constraints = 'python_version>"3.6" and python_version<"3.9"'
+
+all_requirements = sum(
+    [
+        visualization_extras,
+        z3_requirements,
+        [req + "; " + bip_constraints for req in bip_requirements],
```

Comment:
Does docplex have the same version constraints as cplex? It looks like they only ship an sdist and the package metadata says it supports python 3.6, 3.7, and 3.8.

Reply:
Hmm, that's true.  In theory I can redo this as a zip, where we apply the exact constraints to each one.  On the other hand, that would let us partially fulfil an extra on Python 3.6, but not completely.  Probably the intended semantics of `pip install .[all]` are "install as much as possible", so the partial extra is wanted?

Reply:
Also, man, I didn't realise the `sum(..., start=...)` form was Python 3.8!  I got bitten by lazy testing - I only actually tried to pip install from 3.8 and 3.9.

Reply:
I agree, it probably doesn't make sense to install docplex without cplex they're tightly coupled. Although, on 3.6 I actually checked and there is an old release with 3.6 support: https://pypi.org/project/cplex/12.10.0.3/#files. So pip should be smart enough to install that. It's worth testing this before we exclude it on 3.6.

> Also, man, I didn't realise the `sum(..., start=...)` form was Python 3.8! I got bitten by lazy testing - I only actually tried to pip install from 3.8 and 3.9.

Lol, oh well I'm recompiling  3.6 and 3.7 locally (I built them too long ago they're incompatible with my other system lib versions) right now and haven't tested them yet either.

Reply:
The other thing I'm thinking about here is cplex doesn't support 32bit x86 platforms (linux i686 or win32) or linux aarch64. All 3 of which are supported qiskit platforms. Adding environment markers for this would be good too. Although win32 is problematic because they don't give us the right knobs to enable win64 but not win32.

This came up before for symengine which supports all our 64bit platforms but not i686 or win32. The best I could come up with for it was: https://github.com/Qiskit/qiskit-terra/blob/main/requirements.txt#L11 which excludes win32 and win64.